Tests/Realm.Tests/Database/AsyncTests.cs

```diff
@@ -155,5 +158,405 @@ static async Task<long> MeasureTiming(Func<Task> func)
+                Assert.That(person, Is.Not.Null);
+                Assert.That(person.FullName, Is.EqualTo(personName));
+            });
+        }
+
+        [Test]
+        public void AsyncWrite_Func_List()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var peopleQuery = _realm.All<Person>();
+                var personName = "Jonh Wickie";
+                Assert.That(peopleQuery.Count(), Is.EqualTo(0));
+
+                var list = await _realm.WriteAsync(() =>
+                {
+                    var list = new List<Person> { new Person { FullName = personName } };
+                    _realm.Add(list);
+                    return list;
+                });
+
+                Assert.That(peopleQuery.Count(), Is.EqualTo(1));
+                Assert.That(peopleQuery.First().FullName, Is.EqualTo(personName));
+                Assert.That(list, Is.Not.Null);
+                Assert.That(list.Count, Is.EqualTo(1));
+                Assert.That(list.First().FullName, Is.EqualTo(personName));
+            });
+        }
+
+        [Test]
+        public void AsyncWrite_Func_Queryable()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var peopleQuery = _realm.All<Person>();
+                var personName = "Jonh Wickie";
+                Assert.That(peopleQuery.Count(), Is.EqualTo(0));
+
+                var queryable = await _realm.WriteAsync(() =>
+                {
+                    var queryable = new List<Person> { new Person { FullName = personName } }.AsQueryable();
+                    _realm.Add(queryable);
+                    return queryable;
+                });
+
+                Assert.That(peopleQuery.Count(), Is.EqualTo(1));
+                Assert.That(peopleQuery.First().FullName, Is.EqualTo(personName));
+                Assert.That(queryable, Is.Not.Null);
+                Assert.That(queryable.Count, Is.EqualTo(1));
+                Assert.That(queryable.First().FullName, Is.EqualTo(personName));
+            });
+        }
+
+        [Test]
+        public void AsyncBeginWrite_StackReferencedNativeException_Does_Not_Throw()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                using var tokenSource2 = new CancellationTokenSource();
+                CancellationToken ct = tokenSource2.Token;
+                var t = Task.Run(async () =>
+                {
+                    var realm = GetRealm();
+                    var transaction = await realm.BeginWriteAsync();
+                    while (true)
+                    {
+                        await Task.Delay(1000);
+                    }
+                }, ct);
+
+                // give time for the lock to be acquired from the other task
+                await Task.Delay(2000);
+
+                Exception ex = null;
+
+                try
+                {
+                    await _realm.BeginWriteAsync();
```

Comment:
How does this complete? As far as I can tell, the background Realm acquires a write lock first and never releases it - how can we acquire the write lock on the `_realm` instance?

Reply:
1. The cancellation token is useless, so I removed it.
1. The task was running without a sync context. It does now.
1. It completes because they are running on different sync contexts on different Realms. So now that I look at it I'm not even sure that it tests what we thought it does. Again @fealebenpae should help us here.